From d42ac2b7f3cba21b382fb2afe935239c2c778f84 Mon Sep 17 00:00:00 2001
From: Ethan Wei <ethanwei@ami.com>
Date: Thu, 5 Oct 2023 00:48:39 -0700
Subject: [PATCH] Check if IPv4 and IPv4 Default Gateway are in the Same Series
 	modified:   src/ethernet_interface.cpp 	modified:   src/util.cpp 
 modified:   src/util.hpp

---
 src/ethernet_interface.cpp | 16 ++++------------
 src/util.cpp               | 13 +++++++++++++
 src/util.hpp               |  8 ++++++++
 3 files changed, 25 insertions(+), 12 deletions(-)

diff --git a/src/ethernet_interface.cpp b/src/ethernet_interface.cpp
index da4e841..daef139 100644
--- a/src/ethernet_interface.cpp
+++ b/src/ethernet_interface.cpp
@@ -303,6 +303,8 @@ ObjectPath EthernetInterface::ip(IP::Protocol protType, std::string ipaddress,
                                             ip_address::Type::IP4_ADDRESS);
                 if (!ipgateway.empty())
                 {
+                    ip_address::isSameSeries(ipaddress, ipgateway,
+                                             prefixLength);
                     gateway.emplace(
                         stdplus::fromStr<stdplus::In4Addr>(ipgateway));
                     ip_address::isValidIPv4Addr(
@@ -1207,18 +1209,8 @@ std::string EthernetInterface::defaultGateway(std::string gateway)
             if (addr.second->type() == IP::Protocol::IPv4 &&
                 addr.second->origin() != IP::AddressOrigin::LinkLocal)
             {
-                auto prefixLength = addr.second->prefixLength();
-                auto netmask = htobe32(~UINT32_C(0) << (32 - prefixLength));
-                if (((stdplus::fromStr<stdplus::In4Addr>(
-                          addr.second->address()))
-                         .a.s_addr &
-                     netmask) !=
-                    ((stdplus::fromStr<stdplus::In4Addr>(gateway)).a.s_addr &
-                     netmask))
-                {
-                    throw std::logic_error(
-                        "Gateway address and IP address aren't in the same subnet.");
-                }
+                ip_address::isSameSeries(addr.second->address(), gateway,
+                                         addr.second->prefixLength());
                 break;
             }
         }
diff --git a/src/util.cpp b/src/util.cpp
index de751d5..47e6fc0 100644
--- a/src/util.cpp
+++ b/src/util.cpp
@@ -385,6 +385,19 @@ void isValidIPv6Addr(in6_addr* addr, Type type)
         throw std::invalid_argument(strType + " is Unspecified.");
     }
 }
+
+void isSameSeries(std::string ipStr, std::string gwStr, uint8_t prefixLength)
+{
+    auto ip = (stdplus::fromStr<stdplus::In4Addr>(ipStr)).a.s_addr;
+    auto gw = (stdplus::fromStr<stdplus::In4Addr>(gwStr)).a.s_addr;
+    auto netmask = htobe32(~UINT32_C(0) << (32 - prefixLength));
+
+    if ((ip & netmask) != (gw & netmask))
+    {
+        throw std::logic_error(
+            "Gateway address and IP address aren't in the same subnet.");
+    } // if
+}
 } // namespace ip_address
 } // namespace network
 } // namespace phosphor
diff --git a/src/util.hpp b/src/util.hpp
index 6f1fad8..0db4e67 100644
--- a/src/util.hpp
+++ b/src/util.hpp
@@ -57,6 +57,14 @@ void isValidIPv4Addr(in_addr* addr, Type type);
  */
 void isValidIPv6Addr(in6_addr* addr, Type Type);
 
+/** @brief Check if the IPv4 address and default gateway are in the same series
+ *  @param[in] ipAddr - The IPv4 address
+ *  @param[in] gateway - The IPv4 default gateway address
+ *  @param[in] prefixLength - The prefix length of IPv4
+ */
+void isSameSeries(std::string ipAddr, std::string gateway,
+                  uint8_t prefixLength);
+
 } // namespace ip_address
 
 /* @brief converts a sockaddr for the specified address family into
-- 
2.43.0

